chaste_codegen/_linearity_check.py

```diff
@@ -0,0 +1,134 @@
+            for e in expr.args:
+                res = check_expr(e[0], state_var, membrane_voltage_var, state_vars)
+                if result is not None and res != result:
+                    # We have a difference
+                    result = KINDS.NONLINEAR
+                    break
+                result = res
+    elif isinstance(expr, sp.Mul):
+        # Linear iff only 1 linear operand
+        result = KINDS.NONE
+        lin = 0
+        for op in operands:
+            res = check_expr(op, state_var, membrane_voltage_var, state_vars)
+            if res == KINDS.LINEAR:
+                lin += 1
+            elif res == KINDS.NONLINEAR:
+                lin += 2
+            if lin > 1:
+                result = KINDS.NONLINEAR
+                break
+            elif lin == 1:
+                result = KINDS.LINEAR
+    elif isinstance(expr, sp.Add) or isinstance(expr, sp.boolalg.BooleanFunction) or\
+            isinstance(expr, sp.relational.Relational):
+        # linear if any operand linear, and non-linear
+        result = max_kind(state_var, operands)
+    elif isinstance(expr, sp.Pow):
+        if state_var not in expr.free_symbols:
+            result = max_kind(state_var, operands)
+        elif len(expr.args) == 2 and expr.args[1] == -1:  # x/y divide is represented as x * pow(y, -1)
+            # Linear iff only numerator linear
+            result = check_expr(expr.args[0], state_var, membrane_voltage_var, state_vars)
+        else:
+            result = KINDS.NONLINEAR
+    elif isinstance(expr, sp.log) or isinstance(expr, log10) or isinstance(expr, log2) or\
+            isinstance(expr, cg.RealFunction) or\
+            isinstance(expr, sp.functions.elementary.trigonometric.TrigonometricFunction) or\
+            isinstance(expr, sp.functions.elementary.hyperbolic.HyperbolicFunction) or\
+            isinstance(expr, sp.functions.elementary.trigonometric.InverseTrigonometricFunction) or\
+            isinstance(expr, sp.functions.elementary.hyperbolic.InverseHyperbolicFunction):
+        if state_var not in expr.free_symbols:
+            result = check_expr(expr.args[0], state_var, membrane_voltage_var, state_vars)
+        else:
+            result = KINDS.NONLINEAR
+    return result
+
+
+def get_non_linear_state_vars(derivative_equations, membrane_voltage_var, state_vars, printer):
+    """Returns the state vars whos derivative expressions are nont linear"""
+    # return the state var part from the derivative equations where the rhs is not linear
+    return sorted([eq.lhs.args[0] for eq in derivative_equations
+                   if isinstance(eq.lhs, sp.Derivative) and
+                   eq.lhs.args[0] != membrane_voltage_var and
+                   check_expr(eq.rhs, eq.lhs.args[0], membrane_voltage_var, state_vars) != KINDS.LINEAR],
+                  key=lambda s: printer.doprint(s))
+
+
+def derives_eqs_partial_eval_non_linear(y_derivatives, non_linear_state_vars, membrane_voltage_var, state_vars,
```

Comment:
This function name doesn't really enlighten me as to what it's doing! Maybe `subst_deriv_eqs_non_linear_vars` or similar?